[BUG]: fp128 test and feature check are broken in v2.8.x & clang.

### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Compile-time Error

### Component

libcu++

### Describe the bug

CCCL v2.8.x build is broken w/ clang, if fp128 support is enabled on the host, or if it's built with `-D_CCCL_DISABLE_FLOAT128`

https://godbolt.org/z/YrEa9K7oq

There are two issues. 
- First is that fp128 type use leaks into the GPU-side code, which clang does not support
  - Clang currently has no support for fp128 on the GPU. fp128 is OK to use from the host-only code, but the way the tests are structured, fp128 does leak to the GPU side during GPU-side compilation of [libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp](https://github.com/NVIDIA/cccl/blob/e34aa7f7652df658ccf063932db56d61ffc4bb33/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp#L70)  when clang attempts to generate GPU-side code for the `__host__ __device__ fake_main()` function.

- Second problem is that disabling FP128 support is also broken because the function-like feature macros used to check for the feature are not defined if the feature is disabled. 
  - AFAICT, the problem with disabling FP128 has been solved in CCCL v3.x, but it would be great to backport it to the v2 branch, as it's going to be a while until everyone migrates to v3.



### How to Reproduce

Build and run libcu++ tests in CCCL v2.8.5 with clang on a host which has FP128 support enabled.
Observe failure in extended_data_types.pass.cpp

```
third_party/gpus/cccl/v2_8_4/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp:44:36: error: expression requires 128 bit size '__float128' type support, but target 'nvptx64-nvidia-cuda' does not support it
   44 |   __float128 x5 = __float128(3.14) + __float128(3.14);
      |                                    ^
third_party/gpus/cccl/v2_8_4/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp:44:36: error: expression requires 128 bit size '__float128' type support, but target 'nvptx64-nvidia-cuda' does not support it
third_party/gpus/cccl/v2_8_4/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp:45:10: error: 'x5' requires 128 bit size '__float128' type support, but target 'nvptx64-nvidia-cuda' does not support it
   45 |   unused(x5);
      |          ^
third_party/gpus/cccl/v2_8_4/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp:44:14: note: 'x5' defined here
   44 |   __float128 x5 = __float128(3.14) + __float128(3.14);
      |              ^
3 errors generated when compiling for sm_100a.
```

Attempt to disable FP128 tests by building with ` -D_CCCL_DISABLE_FLOAT128`. Observe a different build error: 
```
third_party/gpus/cccl/v2_8_4/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp:43:36: \
  error: function-like macro '_CCCL_HAS_FLOAT128' is not defined

```

### Expected behavior

Disabling FP128 support should not break the build.

### Reproduction link

https://godbolt.org/z/YrEa9K7oq

### Operating System

N/A

### nvidia-smi output

N/A

### NVCC version

n/a. It's a clang-specific problem.

--- a/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
+++ b/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
@@ -93,11 +93,15 @@
 
 #if !defined(CCCL_DISABLE_FLOAT128_SUPPORT) && _CCCL_OS(LINUX) && !_CCCL_ARCH(ARM64)
 #  if (defined(__CUDACC_RTC_FLOAT128__) || defined(__SIZEOF_FLOAT128__) || defined(__FLOAT128__)) /*HOST COMPILERS*/
-#    if _CCCL_CUDA_COMPILER(NVHPC) \
-      || ((_CCCL_CUDA_COMPILER(NVCC) || _CCCL_CUDA_COMPILER(CLANG)) && _CCCL_PTX_ARCH() >= 1000) /*DEVICE CODE*/
+#    if _CCCL_CUDA_COMPILATION() // Only NVCC on architectures at least SM100 supports float128 on device
+#      if _CCCL_CUDA_COMPILER(NVCC) && _CCCL_PTX_ARCH() >= 1000 /*DEVICE CODE*/
+#        undef _CCCL_HAS_FLOAT128
+#        define _CCCL_HAS_FLOAT128() 1
+#      endif // _CCCL_CUDA_COMPILER(NVCC) && _CCCL_PTX_ARCH() >= 1000
+#    else // ^^^ _CCCL_CUDA_COMPILATION() ^^^ / vvv !_CCCL_CUDA_COMPILATION() vvv
 #      undef _CCCL_HAS_FLOAT128
 #      define _CCCL_HAS_FLOAT128() 1
-#    endif // CUDA compiler
+#    endif // !_CCCL_CUDA_COMPILATION()
 #  endif // Host compiler support
 #endif // !CCCL_DISABLE_FLOAT128_SUPPORT && _CCCL_OS(LINUX)
 

